proving/worker.go

```diff
@@ -0,0 +1,193 @@
+package proving
+
+import (
+	"bytes"
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"fmt"
+	"io"
+	"sort"
+	"sync"
+	"unsafe"
+
+	"github.com/spacemeshos/post/oracle"
+	"github.com/spacemeshos/post/shared"
+)
+
+type batch struct {
+	Data    []byte
+	Index   uint64
+	Release func()
+}
+
+type solution struct {
+	Nonce uint
+	Index uint64
+}
+
+// ioWorker is a worker that reads labels from disk and writes them to a batch channel to be processed by the
+// labelWorkers.
+//
+// TODO(mafa): use this as base to replace GranSpecificReader / GranSpecificWriter and the persistence package.
+func ioWorker(ctx context.Context, batchChan chan<- *batch, b uint32, source io.ReadCloser) error {
+	defer close(batchChan)
+	defer source.Close()
+	index := uint64(0)
+
+	batchDataPool := sync.Pool{
+		New: func() any {
+			buf := make([]byte, BlocksPerWorker*b)
+			return &buf
+		},
+	}
+
+	for {
+		data := *(batchDataPool.Get().(*[]byte))
+		n, err := source.Read(data)
+		switch {
+		case err == io.EOF || err == nil:
+			// make sure we don't send partial blocks to the label workers.
+			// TODO(mafa): this silently drops partial blocks and should be handled better.
+			n -= n % int(b)
```

Comment:
I think that's incorrect. The next batch will start with the second part of the labels block. It will offset all subsequent label blocks by `n`.

Reply:
`Read` can return any number of bytes between 0 an `len(data)`. If it returns fewer than `len(data)` bytes it will also return an non-`nil` error. So the only case this can happen here is if at the end of the file there are not enough bytes to fill data completely (`err == io.EOF`).

So `n` is something between 0 and `len(data)` and I have to ensure it is a multiple of `B` otherwise we include a few extra bytes in the last AES invocation that have not been read from `source` but are either zeros or leftover from what was in `data` before we passed it to `source.Read`.

There will be no next batch, since after sending this the `ioWorker` will return on line 63.

Reply:
in the case of `err == nil` `n` is guaranteed to be a multiple of `b` so there will only ever be labels dropped at the end of `source` and at most `b - 1`.

Reply:
👍